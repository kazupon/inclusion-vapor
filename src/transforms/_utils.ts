import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type { RootIRNode as VaporRootIRNode } from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
