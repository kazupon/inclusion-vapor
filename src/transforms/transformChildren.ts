import type { NodeTransform } from './types'
export const transformChildren: NodeTransform = (_node, _context) => {}
