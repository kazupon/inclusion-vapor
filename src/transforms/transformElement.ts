// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformElement.ts

import { createSimpleExpression } from '@vue-vapor/compiler-dom'
import { isVoidTag, extend, makeMap } from '@vue-vapor/shared'
import {
  isSvelteElement,
  convertToSourceLocation,
  // isBuiltInDirective,
  DynamicFlag,
  IRNodeTypes
} from '../ir'
import { isValidHTMLNesting } from '../htmlNesting'
import { EMPTY_EXPRESSION } from './utils'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type {
  IRProps,
  IRProp,
  IRPropsStatic,
  SvelteElement,
  SvelteAttribute,
  SvelteBaseDirective,
  SvelteSpreadAttribute,
  SvelteText,
  SvelteBaseNode
} from '../ir'
import type { NodeTransform } from './types'
import type { TransformContext, DirectiveTransformResult } from '../transforms'

export const isReservedProp: ReturnType<typeof makeMap> = /*#__PURE__*/ makeMap(
  // TODO:
  // the leading comma is intentional so empty string "" is also included
  // ',key,ref,ref_for,ref_key,',
  ''
)

export const transformElement: NodeTransform = (_node, context) => {
  return function postTransformElement() {
    const { node } = context
    if (!isSvelteElement(node)) {
      return
    }

    const { name: tag } = node
    const isComponent = node.type === 'InlineComponent'
    const propsResult = buildProps(node, context as TransformContext<SvelteElement>, isComponent)

    ;(isComponent ? transformComponentElement : transformNativeElement)(
      tag,
      propsResult,
      context as TransformContext<SvelteElement>
    )
  }
}

function transformComponentElement(
  tag: string,
  propsResult: PropsResult,
  context: TransformContext<SvelteElement>
) {
  const asset = true
  context.component.add(tag)
  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  const root = context.root === context.parent && (context.parent.node.children || []).length === 1

  context.registerOperation({
    type: IRNodeTypes.CREATE_COMPONENT_NODE,
    id: context.reference(),
    tag,
    props: propsResult[0] ? propsResult[1] : [propsResult[1]],
    asset,
    root,
    slots: context.slots,
    once: context.inVOnce
  })
  context.slots = []
}

function transformNativeElement(
  tag: string,
  propsResult: PropsResult,
  context: TransformContext<SvelteElement>
) {
  const { scopeId } = context.options

  let template = ''
  template += `<${tag}`
  if (scopeId) {
    template += ` ${scopeId}`
  }

  if (propsResult[0] /* dynamic props */) {
    // TODO:
    // ...
  } else {
    for (const prop of propsResult[1]) {
      const { key, values } = prop
      if (key.isStatic && values.length === 1 && values[0].isStatic) {
        template += ` ${key.content}`
        if (values[0].content) template += `="${values[0].content}"`
      } else {
        context.registerEffect(values, {
          type: IRNodeTypes.SET_PROP,
          element: context.reference(),
          prop
        })
      }
    }
  }

  template += `>` + context.childrenTemplate.join('')
  // TODO remove unnecessary close tag, e.g. if it's the last element of the template
  if (!isVoidTag(tag)) {
    template += `</${tag}>`
  }

  if (
    context.parent &&
    context.parent.node.type === 'Element' &&
    !isValidHTMLNesting(context.parent.node.name as string, tag)
  ) {
    context.reference()
    context.dynamic.template = context.pushTemplate(template)
    context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  } else {
    context.template += template
  }
}

export type PropsResult =
  | [dynamic: true, props: IRProps[], expressions: SimpleExpressionNode[]]
  | [dynamic: false, props: IRPropsStatic]

export function buildProps(
  node: SvelteElement,
  context: TransformContext<SvelteElement>,
  isComponent: boolean
): PropsResult {
  const props = node.attributes
  if (props.length === 0) {
    return [false, []]
  }

  const dynamicArgs: IRProps[] = []
  const dynamicExpr: SimpleExpressionNode[] = []
  let results: DirectiveTransformResult[] = []

  function _pushMergeArg() {
    if (results.length > 0) {
      dynamicArgs.push(dedupeProperties(results))
      results = []
    }
  }

  for (const prop of props) {
    const result = transformProp(prop, node, context)
    if (result) {
      dynamicExpr.push(result.key, result.value)
      if (isComponent) {
        // component props
        // TODO:
      } else {
        // other static props
        results.push(result)
      }
    }
  }

  // has dynamic key or ...
  // TODO:

  const irProps = dedupeProperties(results)
  return [false, irProps]
}

function transformProp(
  prop: SvelteAttribute | SvelteSpreadAttribute | SvelteBaseDirective,
  _node: SvelteElement,
  context: TransformContext<SvelteElement>
): DirectiveTransformResult | void {
  const name = prop.name as string

  // for `SvelteAttribute`
  if (prop.type === 'Attribute') {
    if (isReservedProp(name)) {
      return
    }
    const key = createSimpleExpression(name, true, convertToSourceLocation(prop, name))
    // TODO: should be handled for multiple values
    const value = prop.value[0]
      ? createSimpleExpression(
          (prop.value[0] as SvelteText).data,
          true,
          convertToSourceLocation(
            prop.value[0] as SvelteBaseNode,
            (prop.value[0] as SvelteText).data
          )
        )
      : EMPTY_EXPRESSION
    return {
      key,
      value
    }
  }

  // for `SvelteSpreadAttribute`
  if (prop.type === 'Spread') {
    return
  }

  // for `SvelteBaseDirective`
  const directiveTransform = context.options.directiveTransforms[name]
  if (directiveTransform) {
    // TODO: should tweak directive transform for svelte
    // return directiveTransform(prop, node, context)
  }

  // TODO: should be handled for svelte built-in directives
  // if (!isBuiltInDirective(name)) {
  //   context.directive.add(name)
  //   context.registerOperation({
  //     type: IRNodeTypes.WITH_DIRECTIVE,
  //     element: context.reference(),
  //     dir: prop,
  //     name,
  //   })
  // }
}

// Dedupe props in an object literal.
// Literal duplicated attributes would have been warned during the parse phase,
// however, it's possible to encounter duplicated `onXXX` handlers with different
// modifiers. We also need to merge static and dynamic class / style attributes.
function dedupeProperties(results: DirectiveTransformResult[]): IRProp[] {
  const knownProps: Map<string, IRProp> = new Map()
  const deduped: IRProp[] = []

  for (const result of results) {
    const prop = resolveDirectiveResult(result)
    // dynamic keys are always allowed
    if (!prop.key.isStatic) {
      deduped.push(prop)
      continue
    }
    const name = prop.key.content
    const existing = knownProps.get(name)
    if (existing) {
      if (name === 'style' || name === 'class') {
        mergePropValues(existing, prop)
      }
      // unexpected duplicate, should have emitted error during parse
    } else {
      knownProps.set(name, prop)
      deduped.push(prop)
    }
  }
  return deduped
}

function resolveDirectiveResult(prop: DirectiveTransformResult): IRProp {
  return extend({}, prop, {
    value: undefined,
    values: [prop.value]
  })
}

function mergePropValues(existing: IRProp, incoming: IRProp) {
  const newValues = incoming.values
  existing.values.push(...newValues)
}
