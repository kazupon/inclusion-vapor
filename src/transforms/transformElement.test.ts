import { describe, expect, test } from 'vitest'
import { makeCompile } from './_utils'
import { transformElement } from './transformElement'
import { transformChildren } from './transformChildren'

const compileWithTransform = makeCompile({
  nodeTransforms: [transformChildren, transformElement]
})

describe('Svelete Attribute Node', () => {
  test('native elements', () => {
    const { ir, code, vaporHelpers } = compileWithTransform(
      `<div class="container">
        <div class="header">
          <p style="color: red;">Hello</p>
          <img src="foo.jpg" width="500" height="600">
          <form action="/submit" method="post">
            <input type="text" />
            <input type="submit">
          </form>
        </div>
      </div>`
    )
    expect(code).toMatchSnapshot()
    expect(ir).toMatchSnapshot()
    expect(vaporHelpers).toMatchSnapshot()
  })

  test('components', () => {
    const { ir, code, vaporHelpers } = compileWithTransform(
      `<div class="container">
        <Foo class="foo">
        </Foo>
        <Bar bar="1">
          <Baz />
        </Bar>
      </div>`
    )
    expect(code).toMatchSnapshot()
    expect(ir).toMatchSnapshot()
    expect(vaporHelpers).toMatchSnapshot()
  })
})
