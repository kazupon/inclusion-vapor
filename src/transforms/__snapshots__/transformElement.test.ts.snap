// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Svelte Attribute Node > components 1`] = `
"import { resolveComponent as _resolveComponent, children as _children, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const _component_Baz = _resolveComponent("Baz")
  const _component_Bar = _resolveComponent("Bar")
  const n9 = t0()
  const n0 = n9.firstChild
  const n1 = _children(n9, 0, 0)
  const n3 = n0.nextSibling
  const n4 = _children(n9, 1, 0)
  const n6 = n4.nextSibling
  const n8 = n3.nextSibling
  const n2 = _createComponent(_component_Foo)
  const n5 = _createComponent(_component_Baz)
  const n7 = _createComponent(_component_Bar)
  return [n0, n1, n2, n3, n4, n5, n6, n7, n8, n9]
}"
`;

exports[`Svelte Attribute Node > components 2`] = `
{
  "block": {
    "dynamic": {
      "children": [
        {
          "children": [
            {
              "children": [],
              "flags": 1,
              "id": 0,
            },
            {
              "children": [
                {
                  "children": [],
                  "flags": 1,
                  "id": 1,
                },
              ],
              "flags": 7,
              "id": 2,
            },
            {
              "children": [],
              "flags": 1,
              "id": 3,
            },
            {
              "children": [
                {
                  "children": [],
                  "flags": 1,
                  "id": 4,
                },
                {
                  "children": [],
                  "flags": 7,
                  "id": 5,
                },
                {
                  "children": [],
                  "flags": 1,
                  "id": 6,
                },
              ],
              "flags": 7,
              "id": 7,
            },
            {
              "children": [],
              "flags": 1,
              "id": 8,
            },
          ],
          "flags": 1,
          "id": 9,
          "template": 0,
        },
      ],
      "flags": 1,
    },
    "effect": [],
    "node": {
      "children": [
        {
          "attributes": [
            {
              "end": 22,
              "name": "class",
              "start": 5,
              "type": "Attribute",
              "value": [
                {
                  "data": "container",
                  "end": 21,
                  "raw": "container",
                  "start": 12,
                  "type": "Text",
                },
              ],
            },
          ],
          "children": [
            {
              "data": "
        ",
              "end": 32,
              "raw": "
        ",
              "start": 23,
              "type": "Text",
            },
            {
              "attributes": [
                {
                  "end": 48,
                  "name": "class",
                  "start": 37,
                  "type": "Attribute",
                  "value": [
                    {
                      "data": "foo",
                      "end": 47,
                      "raw": "foo",
                      "start": 44,
                      "type": "Text",
                    },
                  ],
                },
              ],
              "children": [
                {
                  "data": "
        ",
                  "end": 58,
                  "raw": "
        ",
                  "start": 49,
                  "type": "Text",
                },
              ],
              "end": 64,
              "name": "Foo",
              "start": 32,
              "type": "InlineComponent",
            },
            {
              "data": "
        ",
              "end": 73,
              "raw": "
        ",
              "start": 64,
              "type": "Text",
            },
            {
              "attributes": [
                {
                  "end": 85,
                  "name": "bar",
                  "start": 78,
                  "type": "Attribute",
                  "value": [
                    {
                      "data": "1",
                      "end": 84,
                      "raw": "1",
                      "start": 83,
                      "type": "Text",
                    },
                  ],
                },
              ],
              "children": [
                {
                  "data": "
          ",
                  "end": 97,
                  "raw": "
          ",
                  "start": 86,
                  "type": "Text",
                },
                {
                  "attributes": [],
                  "children": [],
                  "end": 104,
                  "name": "Baz",
                  "start": 97,
                  "type": "InlineComponent",
                },
                {
                  "data": "
        ",
                  "end": 113,
                  "raw": "
        ",
                  "start": 104,
                  "type": "Text",
                },
              ],
              "end": 119,
              "name": "Bar",
              "start": 73,
              "type": "InlineComponent",
            },
            {
              "data": "
      ",
              "end": 126,
              "raw": "
      ",
              "start": 119,
              "type": "Text",
            },
          ],
          "end": 132,
          "name": "div",
          "start": 0,
          "type": "Element",
        },
      ],
      "components": [],
      "directives": [],
      "helpers": Set {},
      "source": "<div class="container">
        <Foo class="foo">
        </Foo>
        <Bar bar="1">
          <Baz />
        </Bar>
      </div>",
      "temps": 0,
      "type": 0,
    },
    "operation": [
      {
        "asset": true,
        "id": 2,
        "once": false,
        "props": [
          [],
        ],
        "root": false,
        "slots": [],
        "tag": "Foo",
        "type": 13,
      },
      {
        "asset": true,
        "id": 5,
        "once": false,
        "props": [
          [],
        ],
        "root": false,
        "slots": [],
        "tag": "Baz",
        "type": 13,
      },
      {
        "asset": true,
        "id": 7,
        "once": false,
        "props": [
          [],
        ],
        "root": false,
        "slots": [],
        "tag": "Bar",
        "type": 13,
      },
    ],
    "returns": [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
    ],
    "type": 1,
  },
  "component": Set {
    "Foo",
    "Baz",
    "Bar",
  },
  "directive": Set {},
  "node": {
    "children": [
      {
        "attributes": [
          {
            "end": 22,
            "name": "class",
            "start": 5,
            "type": "Attribute",
            "value": [
              {
                "data": "container",
                "end": 21,
                "raw": "container",
                "start": 12,
                "type": "Text",
              },
            ],
          },
        ],
        "children": [
          {
            "data": "
        ",
            "end": 32,
            "raw": "
        ",
            "start": 23,
            "type": "Text",
          },
          {
            "attributes": [
              {
                "end": 48,
                "name": "class",
                "start": 37,
                "type": "Attribute",
                "value": [
                  {
                    "data": "foo",
                    "end": 47,
                    "raw": "foo",
                    "start": 44,
                    "type": "Text",
                  },
                ],
              },
            ],
            "children": [
              {
                "data": "
        ",
                "end": 58,
                "raw": "
        ",
                "start": 49,
                "type": "Text",
              },
            ],
            "end": 64,
            "name": "Foo",
            "start": 32,
            "type": "InlineComponent",
          },
          {
            "data": "
        ",
            "end": 73,
            "raw": "
        ",
            "start": 64,
            "type": "Text",
          },
          {
            "attributes": [
              {
                "end": 85,
                "name": "bar",
                "start": 78,
                "type": "Attribute",
                "value": [
                  {
                    "data": "1",
                    "end": 84,
                    "raw": "1",
                    "start": 83,
                    "type": "Text",
                  },
                ],
              },
            ],
            "children": [
              {
                "data": "
          ",
                "end": 97,
                "raw": "
          ",
                "start": 86,
                "type": "Text",
              },
              {
                "attributes": [],
                "children": [],
                "end": 104,
                "name": "Baz",
                "start": 97,
                "type": "InlineComponent",
              },
              {
                "data": "
        ",
                "end": 113,
                "raw": "
        ",
                "start": 104,
                "type": "Text",
              },
            ],
            "end": 119,
            "name": "Bar",
            "start": 73,
            "type": "InlineComponent",
          },
          {
            "data": "
      ",
            "end": 126,
            "raw": "
      ",
            "start": 119,
            "type": "Text",
          },
        ],
        "end": 132,
        "name": "div",
        "start": 0,
        "type": "Element",
      },
    ],
    "components": [],
    "directives": [],
    "helpers": Set {},
    "source": "<div class="container">
        <Foo class="foo">
        </Foo>
        <Bar bar="1">
          <Baz />
        </Bar>
      </div>",
    "temps": 0,
    "type": 0,
  },
  "source": "<div class="container">
        <Foo class="foo">
        </Foo>
        <Bar bar="1">
          <Baz />
        </Bar>
      </div>",
  "template": [
    "<div class="container"></div>",
  ],
  "type": 0,
}
`;

exports[`Svelte Attribute Node > components 3`] = `
Set {
  "resolveComponent",
  "children",
  "createComponent",
  "template",
}
`;

exports[`Svelte Attribute Node > native elements 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<p style=\\"color: red;\\"></p>")
const t1 = _template("<img src=\\"foo.jpg\\" width=\\"500\\" height=\\"600\\">")
const t2 = _template("<input type=\\"text\\">")
const t3 = _template("<input type=\\"submit\\">")
const t4 = _template("<form action=\\"/submit\\" method=\\"post\\"></form>")
const t5 = _template("<div class=\\"header\\"></div>")
const t6 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const n16 = t6()
  const n0 = n16.firstChild
  const n14 = n0.nextSibling
  const n14 = t5()
  const n1 = n14.firstChild
  const n3 = n1.nextSibling
  const n3 = t0()
  const n2 = n3.firstChild
  const n4 = n3.nextSibling
  const n5 = n4.nextSibling
  const n5 = t1()
  const n6 = n5.nextSibling
  const n12 = n6.nextSibling
  const n12 = t4()
  const n7 = n12.firstChild
  const n8 = n7.nextSibling
  const n8 = t2()
  const n9 = n8.nextSibling
  const n10 = n9.nextSibling
  const n10 = t3()
  const n11 = n10.nextSibling
  const n13 = n12.nextSibling
  const n15 = n14.nextSibling
  return [n0, n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11, n12, n13, n14, n15, n16]
}"
`;

exports[`Svelte Attribute Node > native elements 2`] = `
{
  "block": {
    "dynamic": {
      "children": [
        {
          "children": [
            {
              "children": [],
              "flags": 1,
              "id": 0,
            },
            {
              "children": [
                {
                  "children": [],
                  "flags": 1,
                  "id": 1,
                },
                {
                  "children": [
                    {
                      "children": [],
                      "flags": 1,
                      "id": 2,
                    },
                  ],
                  "flags": 1,
                  "id": 3,
                  "template": 0,
                },
                {
                  "children": [],
                  "flags": 1,
                  "id": 4,
                },
                {
                  "children": [],
                  "flags": 1,
                  "id": 5,
                  "template": 1,
                },
                {
                  "children": [],
                  "flags": 1,
                  "id": 6,
                },
                {
                  "children": [
                    {
                      "children": [],
                      "flags": 1,
                      "id": 7,
                    },
                    {
                      "children": [],
                      "flags": 1,
                      "id": 8,
                      "template": 2,
                    },
                    {
                      "children": [],
                      "flags": 1,
                      "id": 9,
                    },
                    {
                      "children": [],
                      "flags": 1,
                      "id": 10,
                      "template": 3,
                    },
                    {
                      "children": [],
                      "flags": 1,
                      "id": 11,
                    },
                  ],
                  "flags": 1,
                  "id": 12,
                  "template": 4,
                },
                {
                  "children": [],
                  "flags": 1,
                  "id": 13,
                },
              ],
              "flags": 1,
              "id": 14,
              "template": 5,
            },
            {
              "children": [],
              "flags": 1,
              "id": 15,
            },
          ],
          "flags": 1,
          "id": 16,
          "template": 6,
        },
      ],
      "flags": 1,
    },
    "effect": [],
    "node": {
      "children": [
        {
          "attributes": [
            {
              "end": 22,
              "name": "class",
              "start": 5,
              "type": "Attribute",
              "value": [
                {
                  "data": "container",
                  "end": 21,
                  "raw": "container",
                  "start": 12,
                  "type": "Text",
                },
              ],
            },
          ],
          "children": [
            {
              "data": "
        ",
              "end": 32,
              "raw": "
        ",
              "start": 23,
              "type": "Text",
            },
            {
              "attributes": [
                {
                  "end": 51,
                  "name": "class",
                  "start": 37,
                  "type": "Attribute",
                  "value": [
                    {
                      "data": "header",
                      "end": 50,
                      "raw": "header",
                      "start": 44,
                      "type": "Text",
                    },
                  ],
                },
              ],
              "children": [
                {
                  "data": "
          ",
                  "end": 63,
                  "raw": "
          ",
                  "start": 52,
                  "type": "Text",
                },
                {
                  "attributes": [
                    {
                      "end": 85,
                      "name": "style",
                      "start": 66,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "color: red;",
                          "end": 84,
                          "raw": "color: red;",
                          "start": 73,
                          "type": "Text",
                        },
                      ],
                    },
                  ],
                  "children": [
                    {
                      "data": "Hello",
                      "end": 91,
                      "raw": "Hello",
                      "start": 86,
                      "type": "Text",
                    },
                  ],
                  "end": 95,
                  "name": "p",
                  "start": 63,
                  "type": "Element",
                },
                {
                  "data": "
          ",
                  "end": 106,
                  "raw": "
          ",
                  "start": 95,
                  "type": "Text",
                },
                {
                  "attributes": [
                    {
                      "end": 124,
                      "name": "src",
                      "start": 111,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "foo.jpg",
                          "end": 123,
                          "raw": "foo.jpg",
                          "start": 116,
                          "type": "Text",
                        },
                      ],
                    },
                    {
                      "end": 136,
                      "name": "width",
                      "start": 125,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "500",
                          "end": 135,
                          "raw": "500",
                          "start": 132,
                          "type": "Text",
                        },
                      ],
                    },
                    {
                      "end": 149,
                      "name": "height",
                      "start": 137,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "600",
                          "end": 148,
                          "raw": "600",
                          "start": 145,
                          "type": "Text",
                        },
                      ],
                    },
                  ],
                  "children": [],
                  "end": 150,
                  "name": "img",
                  "start": 106,
                  "type": "Element",
                },
                {
                  "data": "
          ",
                  "end": 161,
                  "raw": "
          ",
                  "start": 150,
                  "type": "Text",
                },
                {
                  "attributes": [
                    {
                      "end": 183,
                      "name": "action",
                      "start": 167,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "/submit",
                          "end": 182,
                          "raw": "/submit",
                          "start": 175,
                          "type": "Text",
                        },
                      ],
                    },
                    {
                      "end": 197,
                      "name": "method",
                      "start": 184,
                      "type": "Attribute",
                      "value": [
                        {
                          "data": "post",
                          "end": 196,
                          "raw": "post",
                          "start": 192,
                          "type": "Text",
                        },
                      ],
                    },
                  ],
                  "children": [
                    {
                      "data": "
            ",
                      "end": 211,
                      "raw": "
            ",
                      "start": 198,
                      "type": "Text",
                    },
                    {
                      "attributes": [
                        {
                          "end": 229,
                          "name": "type",
                          "start": 218,
                          "type": "Attribute",
                          "value": [
                            {
                              "data": "text",
                              "end": 228,
                              "raw": "text",
                              "start": 224,
                              "type": "Text",
                            },
                          ],
                        },
                      ],
                      "children": [],
                      "end": 232,
                      "name": "input",
                      "start": 211,
                      "type": "Element",
                    },
                    {
                      "data": "
            ",
                      "end": 245,
                      "raw": "
            ",
                      "start": 232,
                      "type": "Text",
                    },
                    {
                      "attributes": [
                        {
                          "end": 265,
                          "name": "type",
                          "start": 252,
                          "type": "Attribute",
                          "value": [
                            {
                              "data": "submit",
                              "end": 264,
                              "raw": "submit",
                              "start": 258,
                              "type": "Text",
                            },
                          ],
                        },
                      ],
                      "children": [],
                      "end": 266,
                      "name": "input",
                      "start": 245,
                      "type": "Element",
                    },
                    {
                      "data": "
          ",
                      "end": 277,
                      "raw": "
          ",
                      "start": 266,
                      "type": "Text",
                    },
                  ],
                  "end": 284,
                  "name": "form",
                  "start": 161,
                  "type": "Element",
                },
                {
                  "data": "
        ",
                  "end": 293,
                  "raw": "
        ",
                  "start": 284,
                  "type": "Text",
                },
              ],
              "end": 299,
              "name": "div",
              "start": 32,
              "type": "Element",
            },
            {
              "data": "
      ",
              "end": 306,
              "raw": "
      ",
              "start": 299,
              "type": "Text",
            },
          ],
          "end": 312,
          "name": "div",
          "start": 0,
          "type": "Element",
        },
      ],
      "components": [],
      "directives": [],
      "helpers": Set {},
      "source": "<div class="container">
        <div class="header">
          <p style="color: red;">Hello</p>
          <img src="foo.jpg" width="500" height="600">
          <form action="/submit" method="post">
            <input type="text" />
            <input type="submit">
          </form>
        </div>
      </div>",
      "temps": 0,
      "type": 0,
    },
    "operation": [],
    "returns": [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
    ],
    "type": 1,
  },
  "component": Set {},
  "directive": Set {},
  "node": {
    "children": [
      {
        "attributes": [
          {
            "end": 22,
            "name": "class",
            "start": 5,
            "type": "Attribute",
            "value": [
              {
                "data": "container",
                "end": 21,
                "raw": "container",
                "start": 12,
                "type": "Text",
              },
            ],
          },
        ],
        "children": [
          {
            "data": "
        ",
            "end": 32,
            "raw": "
        ",
            "start": 23,
            "type": "Text",
          },
          {
            "attributes": [
              {
                "end": 51,
                "name": "class",
                "start": 37,
                "type": "Attribute",
                "value": [
                  {
                    "data": "header",
                    "end": 50,
                    "raw": "header",
                    "start": 44,
                    "type": "Text",
                  },
                ],
              },
            ],
            "children": [
              {
                "data": "
          ",
                "end": 63,
                "raw": "
          ",
                "start": 52,
                "type": "Text",
              },
              {
                "attributes": [
                  {
                    "end": 85,
                    "name": "style",
                    "start": 66,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "color: red;",
                        "end": 84,
                        "raw": "color: red;",
                        "start": 73,
                        "type": "Text",
                      },
                    ],
                  },
                ],
                "children": [
                  {
                    "data": "Hello",
                    "end": 91,
                    "raw": "Hello",
                    "start": 86,
                    "type": "Text",
                  },
                ],
                "end": 95,
                "name": "p",
                "start": 63,
                "type": "Element",
              },
              {
                "data": "
          ",
                "end": 106,
                "raw": "
          ",
                "start": 95,
                "type": "Text",
              },
              {
                "attributes": [
                  {
                    "end": 124,
                    "name": "src",
                    "start": 111,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "foo.jpg",
                        "end": 123,
                        "raw": "foo.jpg",
                        "start": 116,
                        "type": "Text",
                      },
                    ],
                  },
                  {
                    "end": 136,
                    "name": "width",
                    "start": 125,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "500",
                        "end": 135,
                        "raw": "500",
                        "start": 132,
                        "type": "Text",
                      },
                    ],
                  },
                  {
                    "end": 149,
                    "name": "height",
                    "start": 137,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "600",
                        "end": 148,
                        "raw": "600",
                        "start": 145,
                        "type": "Text",
                      },
                    ],
                  },
                ],
                "children": [],
                "end": 150,
                "name": "img",
                "start": 106,
                "type": "Element",
              },
              {
                "data": "
          ",
                "end": 161,
                "raw": "
          ",
                "start": 150,
                "type": "Text",
              },
              {
                "attributes": [
                  {
                    "end": 183,
                    "name": "action",
                    "start": 167,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "/submit",
                        "end": 182,
                        "raw": "/submit",
                        "start": 175,
                        "type": "Text",
                      },
                    ],
                  },
                  {
                    "end": 197,
                    "name": "method",
                    "start": 184,
                    "type": "Attribute",
                    "value": [
                      {
                        "data": "post",
                        "end": 196,
                        "raw": "post",
                        "start": 192,
                        "type": "Text",
                      },
                    ],
                  },
                ],
                "children": [
                  {
                    "data": "
            ",
                    "end": 211,
                    "raw": "
            ",
                    "start": 198,
                    "type": "Text",
                  },
                  {
                    "attributes": [
                      {
                        "end": 229,
                        "name": "type",
                        "start": 218,
                        "type": "Attribute",
                        "value": [
                          {
                            "data": "text",
                            "end": 228,
                            "raw": "text",
                            "start": 224,
                            "type": "Text",
                          },
                        ],
                      },
                    ],
                    "children": [],
                    "end": 232,
                    "name": "input",
                    "start": 211,
                    "type": "Element",
                  },
                  {
                    "data": "
            ",
                    "end": 245,
                    "raw": "
            ",
                    "start": 232,
                    "type": "Text",
                  },
                  {
                    "attributes": [
                      {
                        "end": 265,
                        "name": "type",
                        "start": 252,
                        "type": "Attribute",
                        "value": [
                          {
                            "data": "submit",
                            "end": 264,
                            "raw": "submit",
                            "start": 258,
                            "type": "Text",
                          },
                        ],
                      },
                    ],
                    "children": [],
                    "end": 266,
                    "name": "input",
                    "start": 245,
                    "type": "Element",
                  },
                  {
                    "data": "
          ",
                    "end": 277,
                    "raw": "
          ",
                    "start": 266,
                    "type": "Text",
                  },
                ],
                "end": 284,
                "name": "form",
                "start": 161,
                "type": "Element",
              },
              {
                "data": "
        ",
                "end": 293,
                "raw": "
        ",
                "start": 284,
                "type": "Text",
              },
            ],
            "end": 299,
            "name": "div",
            "start": 32,
            "type": "Element",
          },
          {
            "data": "
      ",
            "end": 306,
            "raw": "
      ",
            "start": 299,
            "type": "Text",
          },
        ],
        "end": 312,
        "name": "div",
        "start": 0,
        "type": "Element",
      },
    ],
    "components": [],
    "directives": [],
    "helpers": Set {},
    "source": "<div class="container">
        <div class="header">
          <p style="color: red;">Hello</p>
          <img src="foo.jpg" width="500" height="600">
          <form action="/submit" method="post">
            <input type="text" />
            <input type="submit">
          </form>
        </div>
      </div>",
    "temps": 0,
    "type": 0,
  },
  "source": "<div class="container">
        <div class="header">
          <p style="color: red;">Hello</p>
          <img src="foo.jpg" width="500" height="600">
          <form action="/submit" method="post">
            <input type="text" />
            <input type="submit">
          </form>
        </div>
      </div>",
  "template": [
    "<p style="color: red;"></p>",
    "<img src="foo.jpg" width="500" height="600">",
    "<input type="text">",
    "<input type="submit">",
    "<form action="/submit" method="post"></form>",
    "<div class="header"></div>",
    "<div class="container"></div>",
  ],
  "type": 0,
}
`;

exports[`Svelte Attribute Node > native elements 3`] = `
Set {
  "template",
}
`;
